compiler_opt/rl/corpus.py

```diff
@@ -30,9 +32,45 @@ class ModuleSpec:
   """
   name: str
   exec_cmd: Tuple[str, ...] = ()
+  size: int = 0
 
 
-def build_modulespecs_from_datapath(
+class Corpus:
+  """Represents a corpus. Comes along with some utility functions."""
+
+  def __init__(self,
+               data_path: str,
+               additional_flags: Tuple[str, ...] = (),
+               delete_flags: Tuple[str, ...] = (),
+               module_specs: Optional[List[ModuleSpec]] = None):
+    if module_specs is not None:
+      self._module_specs = module_specs
+    else:
+      self._module_specs = _build_modulespecs_from_datapath(
+          data_path=data_path,
+          additional_flags=additional_flags,
+          delete_flags=delete_flags)
+    self._root_dir = data_path
+    self.size = len(self._module_specs)
+    self._module_specs.sort(key=lambda m: m.size, reverse=True)
+
+  def sample(self, k: int) -> List[ModuleSpec]:
+    """Samples `k` module_specs, sorting by size descending."""
+    k = min(len(self._module_specs), k)
+    if k < 1:
+      raise ValueError('Attempting to sample <1 module specs from corpus.')
+    return list(
+        sorted(
```

Comment:
this is a breaking change tho, no?

but - IIRC, the goal of starting to look at the file size was to enable creating samples where the distribution of the file size in the sample doesn't vary by much - the intuition being that .bc size is a reasonable proxy for compile time, and if samples always have a few gigantic modules, they'll always be ignored.

I don't think the change achieves that, right?

Separately, I'd offer the size-based sampling as an opt-in.

Reply:
> this is a breaking change tho, no?

breaking in the sense that it changes compile order, yes.

> creating samples where the distribution of the file size in the sample doesn't vary by much

I wasn't aware of this, was it intentional to ignore large modules? Do we want to modify the distribution of the sample in some way?

I'll make the sorting optional, then it'll have the same effect as before on training. The sorting was more for the validation data collector (which was consistently being unlucky with a large module at the back)

Reply:
> I wasn't aware of this, was it intentional to ignore large modules? Do we want to modify the distribution of the sample in some way?

Not intentional. So, the data collector is happy to fire N jobs and get a subset back and cancel the rest, in the interest of time. To echo @yundiqian , consider this: say you have N=100. You consistently get 1 module that's enormous. That one would take 5 minutes to compile. Everything else takes 5 seconds.

The "size-based sampling" idea was about having a way we can say "we prefer having similarly sized modules in each sample, if possible". Not sure how this would be formally expressed though.

What's the validation data collector?

Reply:
> "we prefer having similarly sized modules in each sample, if possible"

does "Split the modules by size into N buckets, based on size, in size-order, such that the smallest bucket has the smallest 1/N modules. Sample round-robin from the N buckets, schedule tasks in the order of sampling. " sound about right? Minimizes risk of only having large modules compile first and leaving no time left for tiny modules.

> What's the validation data collector?

Compiles the entirety of a separate-from-training "validation" dataset and returns the performance of the currently trained model on it as the model is being trained. The thing we talked about offline last week

Reply:
the bucketing would work, but not sure if it's sufficiently randomized - @yundiqian?

> validation data collector
Aaa! OK. Got it. We could have that guy do the sorting on its side, or? Anyway, for now, let's do what you have as an opt-in (unless the validation one can do it on its end - haven't seen the impl).

Reply:
this definitely breaks the randomization, but i'm not sure whether it affects the result or not. we can implement it as the additional sampling algorithm so both the random sampling and bucketing sampling are supported

btw, if we do the ordering of the sampled modules here, for the round robin dispatching algorithm, we should probably do 1 2 3 3 2 1 instead of 1 2 3 1 2 3 in dispatching

Reply:
on the 1 2 3 3 2 1 part: you mean to pick the buckets in random order, right?

Reply:
no, i mean doing the 12345543211234554321 in dispatching assuming 20 jobs for 5 workers if the 20 jobs are sorted with descending order in compilation time. this gives you slightly more balanced workload for workers

if you do the current 12345123451234512345, the worker 1 is of the heaviest load and worker 5 is of the lightest load.

Reply:
ah, that might be better fixed with a load-balancing job dispatcher though, since there isn't a guarantee that 123 321 123 is the most optimal scheduling. Right now I have it doing 123 321 213 231, which should be "even" enough (on average, anyway) for the round-robin job dispatcher.